client/lib/track-element-size/index.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * External dependencies
+ */
+import { useRef, useEffect, useState, useCallback } from 'react';
+import { throttle } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import afterLayoutFlush from 'lib/after-layout-flush';
+
+const THROTTLE_RATE = 200;
+
+type UseWindowResizeRect = ReturnType< typeof Element.prototype.getBoundingClientRect >;
+
+// Function type for the expected callback for `useWindowResizeCallback`.
+interface UseWindowResizeCallback {
+	( node: Element | null | undefined ): void;
+}
+
+/**
+ * React hook that subscribes a consumer to changes to the bounding client rect of an element, based
+ * on window resize events.
+ * Does not notify when an element resizes due to reasons other than the window resizing.
+ * Uses throttling on the events, to avoid making changes too often.
+ *
+ * @param callback The function to call back on changes. Takes a single parameter: `boundingClientRect`.
+ *
+ * @returns The ref to be set on the consumer component.
+ */
+export function useWindowResizeCallback(
+	callback: ( boundingClientRect: UseWindowResizeRect ) => any
+) {
+	const lastRect = useRef< UseWindowResizeRect | null >( null );
+	const [ element, setElement ] = useState< Element | null >( null );
```

Comment:
Setting state inside a ref callback is not a good idea: the rerender it causes is always redundant. We can store the element in a ref:
```js
const elementRef = useRef< Element | null >( null );
```
and return it (it has type `React.MutableRefObject< Element | null >`) instead of the `callbackRef`.

The custom typedef for `UseWindowResizeCallback` can be completely avoided. All variables have standard React and DOM types.

Reply:
Thank you for making these changes in your commit! 👍